pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java

```diff
@@ -439,11 +451,65 @@ ClientConfigurationData createClientConfiguration() throws UnsupportedAuthentica
     private PulsarClientImpl createClient(final ClientConfigurationData clientConf, final AuthData clientAuthData,
             final String clientAuthMethod, final int protocolVersion) throws PulsarClientException {
         this.connectionPool = new ProxyConnectionPool(clientConf, service.getWorkerGroup(),
-                () -> new ProxyClientCnx(clientConf, service.getWorkerGroup(), clientAuthRole, clientAuthData,
-                        clientAuthMethod, protocolVersion));
+            () -> new ProxyClientCnx(clientConf, service.getWorkerGroup(), clientAuthRole,
+                this::getOrRefreshClientAuthData, clientAuthMethod, protocolVersion,
+                service.getConfiguration().isForwardAuthorizationCredentials()));
         return new PulsarClientImpl(clientConf, service.getWorkerGroup(), connectionPool, service.getTimer());
     }
 
+    private void updateClientAuthData(AuthData clientData) {
+        this.clientAuthData = clientData;
+        authFutureList.getAndSet(Collections.emptyList())
+            .forEach(future -> future.complete(clientData));
+    }
+
+    private CompletableFuture<AuthData> getOrRefreshClientAuthData() {
```

Comment:
Could you walk me through what this method suppose to do? Thanks

Reply:
When client auth data expired and proxy configured with forward authentication data, then this method requests a new token. This method is thread safe, because one ProxyConnection has multiple ProxyClientCnx, and potentially at same time, they request
auth data.